pkg/kubelet/server.go

```diff
@@ -620,80 +628,179 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+			connectionID = strconv.Itoa(int(stream.Identifier()))
+		case api.StreamTypeData:
+			connectionID = strconv.Itoa(int(stream.Identifier()) - 2)
+		default:
+			connectionID = ""
+		}
+
+		glog.V(5).Infof("Automatically assigning connection ID=%s from stream type=%s, ID=%d", connectionID, streamType, stream.Identifier())
+	}
+	return connectionID
+}
+
+// ServePortForward handles a port forwarding request.
+func ServePortForward(w http.ResponseWriter, req *http.Request, portForwarder PortForwarder, podName string, uid types.UID, idleTimeout time.Duration) {
+	streamChan := make(chan httpstream.Stream, 1)
+
+	glog.V(5).Infof("Upgrading port forward response")
+	upgrader := spdy.NewResponseUpgrader()
+	conn := upgrader.UpgradeResponse(w, req, portForwardStreamReceived(streamChan))
 	if conn == nil {
 		return
 	}
 	defer conn.Close()
-	conn.SetIdleTimeout(s.host.StreamingConnectionIdleTimeout())
 
-	var dataStreamLock sync.Mutex
-	dataStreamChans := make(map[string]chan httpstream.Stream)
+	glog.V(5).Infof("Setting port forwarding streaming connection idle timeout to %v", idleTimeout)
+	conn.SetIdleTimeout(idleTimeout)
 
+	var streamLock sync.Mutex
+	streamChans := make(map[string]chan httpstream.Stream)
+
+	glog.V(5).Infof("Waiting for port forward streams")
 Loop:
 	for {
 		select {
 		case <-conn.CloseChan():
+			glog.V(5).Infof("Upgraded connection closed")
 			break Loop
 		case stream := <-streamChan:
-			streamType := stream.Headers().Get(api.StreamType)
-			port := stream.Headers().Get(api.PortHeader)
-			dataStreamLock.Lock()
-			switch streamType {
-			case "error":
-				ch := make(chan httpstream.Stream)
-				dataStreamChans[port] = ch
-				go waitForPortForwardDataStreamAndRun(kubecontainer.GetPodFullName(pod), uid, stream, ch, s.host)
-			case "data":
-				ch, ok := dataStreamChans[port]
-				if ok {
-					ch <- stream
-					delete(dataStreamChans, port)
-				} else {
-					glog.Errorf("Unable to locate data stream channel for port %s", port)
-				}
-			default:
-				glog.Errorf("streamType header must be 'error' or 'data', got: '%s'", streamType)
-				stream.Reset()
+			connectionID := getPortForwardStreamConnectionID(stream)
+			glog.V(5).Infof("(%s) Received new stream", connectionID)
+
+			streamLock.Lock()
+
+			ch, ok := streamChans[connectionID]
+			if !ok {
+				glog.V(5).Infof("(%s) connection ID not in streamChans", connectionID)
+
+				glog.V(5).Infof("(%s) creating stream channel", connectionID)
+				ch = make(chan httpstream.Stream, 2)
+				streamChans[connectionID] = ch
+
+				glog.V(5).Infof("(%s) starting handlePortForwardStreams", connectionID)
+				go func() {
+					handlePortForwardStreams(podName, uid, ch, portForwarder, connectionID)
+					streamLock.Lock()
+					defer streamLock.Unlock()
+					delete(streamChans, connectionID)
+					close(ch)
+				}()
 			}
-			dataStreamLock.Unlock()
+
+			streamLock.Unlock()
+
+			glog.V(5).Infof("(%s) sending stream to stream channel", connectionID)
+			ch <- stream
```

Comment:
So I think I see several problems with this with misbehaving or malicious clients that could DDoS or outright crash the server.

1. A client could create 3 error streams. The first one is fine. The second one comes in and is written to the buffered channel here. The third one comes in and is blocked on the buffered write. Now the second one is received in `handlePortForwardStreams`, causes an error and an early return, then the channel is closed. Now the write of the third stream to the channel panics.

2. A client creates a data and error stream like normal. `handlePortForwardStreams` is no longer consuming from this channel, so future streams created just build up and leave a lot of goroutines blocked to writing on this channel. These built up goroutines will probably all blow up when `handlePortForwardStreams` returns and closes the channel.